koku/masu/processor/ocp/ocp_cost_model_cost_updater.py

```diff
@@ -282,6 +400,82 @@ def _update_monthly_tag_based_cost(self, start_date, end_date):
+                        metric_constants.SUPPLEMENTARY_COST_TYPE,
+                    ]:
+                        if rate_type == metric_constants.INFRASTRUCTURE_COST_TYPE:
+                            if not self._tag_infra_rates.get(rate_term, {}):
+                                continue
+                            rate_dict = self._tag_infra_rates.get(rate_term)
+                            default_rate_dict = None
+                            if self._tag_default_infra_rates.get(rate_term):
+                                default_rate_dict = self._tag_default_infra_rates.get(rate_term)
+                        else:
+                            if not self._tag_supplementary_rates.get(rate_term, {}):
+                                continue
+                            rate_dict = self._tag_supplementary_rates.get(rate_term)
+                            default_rate_dict = None
+                            if self._tag_default_supplementary_rates.get(rate_term):
+                                default_rate_dict = self._tag_default_supplementary_rates.get(rate_term)
+
+                        LOG.info(
+                            "Updating tag based monthly %s cost for"
+                            "\n\tSchema: %s \n\t%s Provider: %s (%s) \n\tDates: %s - %s",
+                            cost_type,
+                            self._schema,
+                            self._provider.type,
+                            self._provider.name,
+                            self._provider_uuid,
+                            start_date,
+                            end_date,
+                        )
+
+                        if cost_type == "Node":
+                            key_rate_dict = self._build_node_tag_rate_case_statement_str(
+                                rate_dict, start_date, default_rate_dict
+                            )
+                            unallocated_key_rate_dict = self._build_node_tag_rate_case_statement_str(
+                                rate_dict, start_date, default_rate_dict, unallocated=True
+                            )
+                            labels = self._build_labels_case_statement_str(
+                                rate_dict, "pod_labels", default_rate=default_rate_dict
+                            )
+                        elif cost_type == "PVC":
+                            key_rate_dict = self._build_volume_tag_rate_case_statement_str(
+                                rate_dict, start_date, default_rate_dict
+                            )
+                            labels = self._build_labels_case_statement_str(
+                                rate_dict, "volume_labels", default_rate=default_rate_dict
+                            )
+
+                        for tag_key, case_statements in key_rate_dict.items():
+                            unallocated_case_statements = unallocated_key_rate_dict.get(tag_key, {})
```

Comment:
Will `cost_type` always be `PVC` or `Node`? It looks like there is also `Cluster`, so how does this `for` loop not throw a `variable referenced before assignment` error when iterating over the `key_rate_dict.items()` when the cost_type is `Cluster`?

Also, is `unallocated_key_rate_dict` defined anywhere other than if the cost_type is `Node`?

Reply:
Oh I see whats happening here. The first cost_type in `OCPUsageLineItemDailySummary.MONTHLY_COST_RATE_MAP` is `Node`, so `key_rate_dict` and `unallocated_key_rate_dict` are defined the first time thru the `for` loop and then re-used the next time through. Is this intentional?

Reply:
Prolly not, I shall fix.